File: leetcode/binary-tree-postorder-traversal/binary-tree-postorder-traversal-iter.rb
# LeetCode #145 - Binary Tree Postorder Traversal
# https://leetcode.com/problems/binary-tree-postorder-traversal/
# Iterative solution.

# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer[]}
def postorder_traversal(root)
  out = []
  stack = []
  last = nil

  while root || !stack.empty?
    # Go left as far as possible.
    while root
      stack << root
      root = root.left
    end

    cur = stack.last

    if cur.right && cur.right != last
      # We can go right from here but we haven't. Do that next.
      root = cur.right
    else
      # We can't go right, or already did. Visit the current node and retreat.
      out << cur.val
      last = cur
      stack.pop
    end
  end

  out
end

File: leetcode/binary-tree-postorder-traversal/binary-tree-postorder-traversal-iter.java
// LeetCode #145 - Binary Tree Postorder Traversal
// https://leetcode.com/problems/binary-tree-postorder-traversal/
// Iterative implementation.

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> values = new ArrayList<>();
        postorder(root, values::add);
        return values;
    }

    private static void postorder(TreeNode root, IntConsumer action) {
        var stack = Collections.asLifoQueue(new ArrayDeque<TreeNode>());
        TreeNode last = null;

        while (root != null || !stack.isEmpty()) {
            // Go left as far as possible.
            for (; root != null; root = root.left) stack.add(root);

            var cur = stack.element();

            if (cur.right != null && cur.right != last) {
                // We can go right but haven't, so do that next.
                root = cur.right;
            } else {
                // Do the postorder action and retreat.
                action.accept(cur.val);
                last = cur;
                stack.remove();
            }
        }
    }
}
